lib/tasks/BackbeatTask.js

```diff
@@ -0,0 +1,75 @@
+const BackOff = require('backo');
+const joi = require('joi');
+
+const BACKOFF_PARAMS = { min: 1000, max: 300000, jitter: 0.1, factor: 1.5 };
+
+const backbeatTaskConfigJoi = {
+    retryTimeoutS: joi.number().default(300),
+};
+
+/**
+ * @class BackbeatTask
+ */
+class BackbeatTask {
+    /**
+     * @constructor
+     * @param {Object} [config] - config object containing various
+     *   task-specific configuration
+     * @param {Number} [config.retryTimeoutS] - timeout of retries, in
+     *   seconds
+     */
+    constructor(config) {
+        this.config = joi.attempt(config || {}, backbeatTaskConfigJoi,
+                                  'invalid backbeat task config');
+    }
+
+    retry(args, done) {
+        const { actionDesc, logFields,
+                actionFunc, shouldRetryFunc, onRetryFunc, log } = args;
+        const backoffCtx = new BackOff(BACKOFF_PARAMS);
+        let nbRetries = 0;
+        const startTime = Date.now();
+        const self = this;
+
+        function _handleRes(...args) {
+            const err = args[0];
+            if (!err) {
+                if (nbRetries > 0) {
+                    const now = Date.now();
+                    log.info(`succeeded to ${actionDesc} after retries`,
+                             Object.assign({
+                                 nbRetries,
+                                 retryTotalMs: `${now - startTime}`,
+                             }, logFields || {}));
+                }
+                return done(...args);
+            }
+            if (!shouldRetryFunc(err)) {
+                return done(err);
+            }
+            if (onRetryFunc) {
+                onRetryFunc(err);
```

Comment:
Should this not be returned?

Reply:
no, this is just a custom helper set by the caller to do some side work, we still want to continue processing the failure.